### ISSUE TITLE
version id scheme 

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/chipper#560]
The version id for PhET sims has (imo) gotten a bit confused with the introduction of PhET-iO.  The 'phetio' brand identifier has been tacked on in a non-general way, and I have yet to see anyone provide a coherent/comprehensive description of the version id's syntax and semantics (`getVersionForBrand.js` included).

### Current version id scheme

Here is the current format for public, dev and RC versions respectively, when brand is 'phet':

1.1.0
1.1.0-dev.1
1.1.0-rc.1 

So far, so good. Now here is the current format when brand is 'phetio':

1.1.0-phetio
1.1.0-phetiodev.1
1.1.0-phetiorc.1 

In the public version case, the build process adds a '-phetio' prefix.  In the dev and RC cases, the build process reads the version id from package.json and replaces 'dev' and 'rc' with 'phetiodev' and 'phetiorc' respectively.  

A few problems with this scheme:

(1) It modifies what the developer puts in package.json.
(2) It modifies test and public versions differently.
(3) It combines the "test type" and "brand" info, making it more difficult to parse.
(4) It is not general enough to support future test types or brands.
(5) It's difficult to write a "general form" description.

I'm proposing that we revised the version id scheme, as described below.

### Proposed version id scheme

The version id scheme will consists of 3 chunks:

`number[-test][-brand]`

where:
 
• `number` describes the version number
• `test` describes the test (absence indicates a public version that has been QA tested)
•`brand` describes the brand (absence indicates 'phet' brand)
• square brackets indicate an optional chunk
• hypen is the separator between chunks

Expanding the chunks, we have:

`major.minor.maintenance[-testType.testNumber][-brandName]`

where:

• `major` = major number, sequential integer
• `minor` = minor number, sequential integer, resets to 0 when `major` is changed
• `maintenance` = maintenance number, sequential integer, resets to 0 when `minor` is changed
• `typeType` = the type of test, currently 'dev' or 'rc'
• `testNumber` = sequential integer indicating the test number
• `brandName` = the brand name, e.g. 'phetio'.  Absence indicates 'phet' brand.

The developer specifies the `number[-test]` portion of the version identifier in package.json. The build process add the `[-brandName]` portion, if applicable, to each supported brand that is built.

Examples:

1.5.0 = public version of 'phet' brand
1.5.0-phetio = public version of 'phetio' brand 
1.5.0-x = public version of brand 'x', where 'x' is a client or some future brand
1.5.0-dev.1 = dev test for 'phet' brand
1.5.0-dev.1-phetio = dev test for 'phetio' brand
1.5.0-dev.1-x = dev test for brand 'x'
1.5.0-rc.1 = RC test for 'phet' brand
1.5.0-rc.1-phetio = RC test for 'phetio' brand
1.5.0-rc.1-x = RC test for brand 'x'

Another advantage of this scheme will become apparent if/when we publish all supported brands for a sim simultaneously. We would want all related files to live in the same directory, with version ids that are clearly similar.  For example, for a sim with supported brands 'phet', 'phetio' and 'x':

public versions, all located in 1.0.0/ directory:
1.0.0
1.0.0-phetio
1.0.0-x

dev versions, all located in 1.0.0-dev.15/ directory:
1.0.0-dev.15
1.0.0-dev.15-phetio
1.0.0-dev.15-x

RC version, all located in 1.0.0-rc.1/ directory:
1.0.0-rc.1
1.0.0-rc.1-phetio
1.0.0-rc.1-x

Labeling for discussion at developer meeting.

EDIT (@samreid): fixed a typo

**Comment by samreid** [comment]
Prior discussion about how to name the versions is in https://github.com/phetsims/phet-io/issues/623 and it is codified in chipper/js/getVersionForBrand.js

**Comment by pixelzoom** [comment]
The current scheme of inserting "phetio" in front of "dev" and "rc" in the version identifier breaks the detection scheme that we approved in https://github.com/phetsims/joist/issues/406#issuecomment-284893658.

**Comment by pixelzoom** [comment]
For a summary of other problems with the current version id scheme, see https://github.com/phetsims/joist/issues/411 (clean up version identifier).

**Comment by pixelzoom** [comment]
Here's another possible general form that is closer to what we have now.  The brand chunk comes before the test chunk:

```
number[-brand][-test]
```

E.g.:

1.0.0
1.0.0-phetio
1.0.0-rc.1
1.0.0-phetio-rc.1
1.0.0-dev.1
1.0.0-phetio-dev.1

**Comment by pixelzoom** [comment]
Here's another possible general form that is closer to what we have now.  The brand chunk comes before the test chunk:

```
number[-brand][-test]
```

E.g.:

1.0.0
1.0.0-phetio
1.0.0-rc.1
1.0.0-phetio-rc.1
1.0.0-dev.1
1.0.0-phetio-dev.1

This would probably require fewer changes to the build process.  But imo it's preferable to put the brand chunk last, since it's absence indicates an implicit value ('phet').

**Comment by samreid** [comment]
It appears that the semantic versioning rules (which we looked at previously) does support hyphens in pre-release version identifiers.  http://semver.org/#spec-item-9 . However, I'm not sure that we should be using "pre-release" style to identify production versions like "1.0.0-phetio".

**Comment by samreid** [comment]
Brainstorming: what if instead of changing the version name for phet-io sims, we change the project name.  
Currently:
faradays-law 1.0.0-phetio

Proposed:
faradays-law-phetio 1.0.0

This idea was prompted by on the semantic versioning link above.

**Comment by samreid** [comment]
>  ![image](https://cloud.githubusercontent.com/assets/679486/25013820/593675ba-2032-11e7-8911-d4ea814e9fab.png)

They are different applications, so doesn't it make sense that they would have different names?

**Comment by pixelzoom** [comment]
They are not different applications. They are the same application with different branding and different features enabled.

**Comment by samreid** [comment]
> They are not different applications. They are the same application with different branding and different features enabled.

They also have different code (phet-io sim versions include phetio.js internally and wrappers externally).

**Comment by pixelzoom** [comment]
So... phetio is phet with additional features (and code for those features) added, right?

**Comment by samreid** [comment]
Correct, there is added code and different brand metadata.

**Comment by samreid** [comment]
Would it be simpler from an engineering standpoint if we only had one file (which supports both phet and phet-io)?  If so, we would need @kathy-phet to approve this before moving forward.

**Comment by samreid** [comment]
We are discussing making phet-io development part of the normal process and unifying the version numbers (and branch names) and having the build process build both artifacts.

**Comment by samreid** [comment]
Hence we would take the brand out of the version identifier and branch names.

@jonathanolson and @zepumph and @samreid will work together to investigate how to make this happen.

**Comment by samreid** [comment]
As part of https://github.com/phetsims/chipper/issues/567 we will investigate removing brand from version and branch.

**Comment by pixelzoom** [comment]
In 9/18/17 Slack discussion (developer channel), @jonathanolson needed to publish a version of pendulum lab from a branch named "ubc-study" (see https://github.com/phetsims/pendulum-lab/issues/193).  What should the version id have been?  There was much discussion and no consensus.  The current version id scheme doesn't support something like "ucb-study-dev.1".  How does the need to publish from branches that aren't named "major.minor" (e.g. "1.2") after the version id scheme?

**Comment by pixelzoom** [comment]
11/6/17 dev meeting notes

Current dev directory structure:

```
faradays-law/
  // brand=phet
  1.0.0-dev.1/
    faradays-law_en.html
  1.0.0-rc.1/
    faradays-law_en.html
  1.0.0/
    faradays-law_en.html

  // brand=phet-io
  1.0.0-phetiodev.1/
    faradays-law_en-phetio.html 

  // one offs
  1.0.1-sonification.1/
    faradays-law_en.html   
  1.3.7-phetiodevtest.1/
    faradays-law_en-phetio.html
```   

Desired dev directory structure:

```
faradays-law/
  // brand=phet
  1.0.0-dev.1/
    faradays-law_en.html
  1.0.0-rc.1/
    faradays-law_en.html
  1.0.0/
    faradays-law_en.html

  // brand=phetio

```

Decisions:
• Hypen not allowed in brand names. Change "phet-io" to "phetio".  Typically use camelcase for other brands, e.g. "adaptedFromPhET" or "widgetsInc".

**Comment by mattpen** [comment]
Will the new versioning scheme be an automated step or still be a manual process?

If it is automated should it go in chipper (meaning old versioning scheme still used for some sims) or the build-server (would not be used for dev deploys)?

If we want to include this in perennial, maybe it would be a good idea to migrate devDeploy to the build-server?

**Comment by samreid** [comment]
The heart of chipper 2.0 is automating things to prevent manual errors & labor, so I imagine we would want to codify and automate the versioning too.

**Comment by jonathanolson** [comment]
I'd also be interested in whether we would want to mark debug versions (with assertions included and potentially enabled, see https://github.com/phetsims/phet-io/issues/1149) with 'debug' information in its version.

Proposed version scheme doesn't have ways to add on any "in-the-future" additional tags, like `-debug`.

Thoughts?

**Comment by jonathanolson** [comment]
Or, since we're also not including the locale (affects the build) in the version, presumably we wouldn't include whether it's a debug build?

**Comment by samreid** [comment]
How does the locale get specified?

**Comment by jonathanolson** [comment]
I guess I'm trying to figure out conceptually what the "version" would include. Yotta and such cares a lot about what locale it is (and it may matter whether it's a debug build), and some things care about our build timestamp (which is how we can tell translation updates apart).

Additionally, it seems the version should also include a "one-off" identifier field, so that the directory/etc. can be fully determined from the "version".

**Comment by zepumph** [comment]
Currently is the only thing that specifies the locale in a built sim the html file? If so, and yotta doesn't mind managing locales this way, perhaps we could get away with yotta handling debug mode in the same way, with out touching the version? 

Conversely maybe it would be better to maintain specificity on the version/directory level rather than just with different html files. It also seems like we may want consistency between built versions based on all of these factors (one-off/debug/locale/etc).

**Comment by samreid** [comment]
> Conversely maybe it would be better to maintain specificity on the version/directory level rather than just with different html files. 

It seems our choices are roughly:

faradays-law/1.0.0/faradays-law_en.html
faradays-law/1.0.0-debug/faradays-law_en.html

vs

faradays-law/1.0.0/faradays-law_en.html
faradays-law/1.0.0/faradays-law_en-debug.html

I was envisioning the second option, that the debug version would be supplied with each build.

**Comment by jonathanolson** [comment]
I was also thinking of the second version.

Would we have `faradays-law_en-phetio-sonification-debug.html`? We'd want to decide on an order to the flags.

**Comment by pixelzoom** [comment]
Note to self: review comments on this issue that occurred since 11/17/17.

**Comment by jbphet** [comment]
Personally, I'd prefer version ID instead of structured directories.  The proposal for the change to version ID as I understand it is:
```
major.minor.maintenance[-test]-brand[-oneOffName][-debug]
```
We should decide if `-debug` is the only thing that can appear in this spot or if, like "-test" and "-oneOffName" it can take on different values.  For now, my understanding is that `-debug` is the only proposed value.  If it can take on more values, we will need to figure out how to distinguish it from other values, such as the oneOffName.

This doesn't seem too objectionable to me, especially since the `-debug` versions will rarely if ever be provided to users outside of PhET (I think).  But it is starting to feel like we are trying to cram an awful lot of information into the version ID, and IMO if anything else comes up we should take a pause and investigate some alternatives, such as having some easily decipherable metadata embedded at the top of the built file with the additional information that we need.

**Comment by zepumph** [comment]
I think that the proposal to version ID is
```
major.minor.maintenance[-{{test}}]-brand[-{{oneOffName}}][-debug]
```
Where {{}} represents a template that changes, but `-debug` is just that string. I think this is what @jbphet was saying, but I want to be explicit.

There are many hyphens in this schema. Is it possible that an underscore would help differentiate components and decrease parsing complexity? Maybe something like

```
major.minor.maintenance[-{{test}}]-brand[-{{oneOffName}}][_debug]
```

>the -debug versions will rarely if ever be provided to users outside of PhET (I think)

Developing these debug versions with assertions and such in them is for phet-io clients during their development as well.

**Comment by pixelzoom** [comment]
Personally, I'm loosing the ability to comprehend PhET's versioning and publishing needs. They have become way too complicated, and should be someone's primary responsibility (as I suggested long ago in https://github.com/phetsims/chipper/issues/567#issuecomment-298974310).   So I'm afraid that at this point I don't have any productive thoughts on this issue.

**Comment by jonathanolson** [comment]
A few more notes:

- > "-phet" is stripped out of directory and file names when production versions are deployed to the production server. (AP, KP: We will not change the names on the production server, too problematic.)

  This sounds like it may bite us later if we just search for that string (may match other parts). If we have a one-off name starting with `phet`, does it just get mangled horribly? Or do we need to fully parse filenames out?

- Versioning proposal means files under /build/ get changed, so phetmarks and all of our tooling that expects build/${sim}_en.html to exist needs to change.
- If you check out an older snapshot of a sim (maintenance release), that tooling (including phetmarks) wouldn't be compatible.
- The above will require a decent chunk of Yotta changes, which I'm not excited about (but is fine).

**Comment by samreid** [comment]
Things get a little simpler if we assume that a debug version always gets built, and all supported brands are always built.  Then we don't need to include those attributes in the version (folder) name.

**Comment by ariel-phet** [comment]
Desired dev directory structure:

```
faradays-law/

  // brand=phet
  1.0.0-dev.1-phet/
    faradays-law_en-phet.html
  1.0.0-rc.1-phet/
    faradays-law_en-phet.html
  1.0.0-phet/
    faradays-law_en-phet.html

  // brand=phetio
  1.0.0-dev.1-phetio/
    faradays-law_en-phetio.html
  1.0.0-rc.1-phetio/
    faradays-law_en-phetio.html
  1.0.0-phetio/
    faradays-law_en-phetio.html

  // one off, brand=phetio
  1.1.0-dev.1-phetio-sonification/
    faradays-law_en-phetio-sonification.html
  1.1.0-dev.2-phetio-sonification/
    faradays-law_en-phetio-sonification.html
  1.1.0-phetio-sonification/
    faradays-law_en-phetio-sonification.html

  // one off, brand=phet
  1.1.0-dev.1-phet-foobar/
    faradays-law_en-phet-foobar.html
  1.1.0-dev.2-phet-foobar/
    faradays-law_en-phet-foobar.html
  1.1.0-phetio-sonification/
    faradays-law_en-phet-foobar.html
```

**Comment by ariel-phet** [comment]
Consensus (for the moment) at 11/30/2017 dev meeting, plan on using `_all.html` for all debug versions.

**Comment by jonathanolson** [comment]
> faradays-law_en-phetio.html

I thought from another issue that for phet-io we would be using a version with all locales? (Could omit the `_en`, or if we are enforcing consistency, `faradays-law_all-phetio.html` would be the main phet-io built sim)

**Comment by zepumph** [comment]
I agree with you. I am about 80% sure that we should be building `*_all-phetio.html` always, and never `*_en-phetio.html`. https://github.com/phetsims/chipper/issues/632#issuecomment-348325508 is meant to solidify that. I will update the above "Master Version Schema" above with those results.

**Comment by zepumph** [comment]
As per @ariel-phet's request. I updated the above master schema with https://github.com/phetsims/chipper/issues/632#issuecomment-348329508. If things change again I will comment before updating it.

**Comment by jonathanolson** [comment]
Sounds like we're really sticking to a template if we're including redundant information (e.g. "all" and "debug", "all" and "phet-io", etc.?)

**Comment by pixelzoom** [comment]
12/4/17 dev meeting notes.

In local working copy and build server, support existence of artifacts for multiple brand in `build/` directory.  Directory structure (e.g.):

```
faradays-law/
  build/
    phet/
      faradays-law_en-phet.html
      ...
    phetio/
      faradays-law_all-phetio.html
      ...
```

**Comment by pixelzoom** [comment]
12/4/17 dev meeting notes continued:



*** local ***
  
build/
  phet/
      faradays-law_en-phet.html
      faradays-law_all-phet-debug.html
      faradays-law_all-phet.html
  phetio/
      faradays-law_all-phetio.html
      faradays-law_all-phetio-debug.html


*** spot (current chipper:2.0) ***

1.0.0-dev.1-phet/
      faradays-law_en-phet.html
      faradays-law_all-phet-debug.html
      faradays-law_all-phet.html
1.0.0-dev.1-phetio/
      faradays-law_all-phetio.html
      faradays-law_all-phetio-debug.html


*** spot (proposed chipper:2.0) ***

1.0.0-dev.1/
   phet/
      faradays-law_en-phet.html
      faradays-law_all-phet-debug.html
      faradays-law_all-phet.html
   phetio/
      faradays-law_all-phetio.html
      faradays-law_all-phetio-debug.html

PROS:
• structure matches build/
• easier to transfer to spot
• SR: pulls brand out of top-level directory name

CONS:
• more difficult to “find latest” of a specific brand

CONSENSUS: go with “proposed” above for spot

**Comment by pixelzoom** [comment]
12/4/17 revisions to https://github.com/phetsims/chipper/issues/560#issuecomment-348304508 based on decision in https://github.com/phetsims/chipper/issues/560#issuecomment-349051862:

**dev directory structure** (assuming `brands=phet,phetio`)

```
faradays-law/

  1.0.0-dev.1/
    phet/
      faradays-law_en-phet.html
      faradays-law_all-phet-debug.html
      faradays-law_all-phet.html
    phetio/
      faradays-law_all-phetio.html
      faradays-law_all-phetio-debug.html

  1.0.0-rc.1/
    phet/
      faradays-law_en-phet.html
      faradays-law_all-phet-debug.html
      faradays-law_all-phet.html
    phetio/
      faradays-law_all-phetio.html
      faradays-law_all-phetio-debug.html

  1.0.0/
    phet/
      faradays-law_en-phet.html
      faradays-law_all-phet-debug.html
      faradays-law_all-phet.html
    phetio/
      faradays-law_all-phetio.html
      faradays-law_all-phetio-debug.html

  // one-off is in the directory name
  1.0.1-dev.1-sonification/
    phet/
      faradays-law_en-phet-sonification.html
      faradays-law_all-phet-sonification.html
      faradays-law_all-phet-sonification-debug.html
    phetio/
      faradays-law_all-phetio-sonification.html
      faradays-law_all-phetio-sonification-debug.html
```

**Comment by jonathanolson** [comment]
Example of dev deployment (chipper 2.0 branch with perennial `grunt dev --repo=chains --brands=phet,phet-io`) with both brands (thanks @zepumph for instrumenting chains!!): https://www.colorado.edu/physics/phet/dev/html/chains/1.7.0-dev.9/

There's a lot of path-specific code in wrappers, and it I wasn't able to quickly (10mins) figure out how to get the wrappers to point to the correct sim (without breaking require.js), so I'll leave this for a phet-io person (assigning @zepumph initially).

Additionally, I'll need to coordinate with @mattpen for the build-server work to be compatible with these changes, assigning.

**Comment by jonathanolson** [comment]
It looks kind of funny with the directory names (below the version directory) being `phet`, `phetio` and `adaptedFromPhet` when it would be cleaner to use the raw brand name (`phet`, `phet-io`, `adapted-from-phet`).

Does anyone else feel the same way, or is there a reason why we chose to shorten to "phetio"?

**Comment by zepumph** [comment]
I think in general, phet-io has pushed for hyphens as much as the code will allow (sometimes you need to squash for var names).  I think I'm in favor of seeing the directory "phet-io"

**Comment by pixelzoom** [comment]
@jonathanolson said:
> ... when it would be cleaner to use the raw brand name (`phet`, `phet-io`, `adapted-from-phet`).

What is a "raw brand name"? I thought we decided that the brand names were changing to camelcase, and the hyphens and punctuation were hereafter verboten in brand names.

**Comment by jonathanolson** [comment]
> What is a "raw brand name"? I thought we decided that the brand names were changing to camelcase, and the hyphens and punctuation were hereafter verboten in brand names.

I was not aware we were changing the brand names, but instead mapping them differently for the filenames, like:

> Change brand "phet-io" to "phetio". Typically use camelcase for other brands, e.g. "adaptedFromPhET" or "widgetsInc".

Instead, it sounds like a more major change, where we need to refactor a lot of common/sim code, change query parameters (`brand=phetio`), ensure backwards compatibility in build-server/yotta (but presumably don't treat phetio and phet-io as two different brands). I assume we still live with phet-io.colorado.edu (sounds like too much effort to change), but do we change repository names that start with `phet-io`?

I don't recall discussing any of those trade-offs, or scope of any renaming that was past "how the filenames and directory names are", so I'd like to get confirmation from other developers (or @ariel-phet ) if this is actually the intended goal (as it will deserve its own issues for handling a change that large).

**Comment by jonathanolson** [comment]
If it wasn't clear, my opinion is that usage with the hyphen (`phet-io`) is pervasive in code and names, and the costs of trying to change this is nowhere near the benefit. Instead, the remapping for filenames (but not directories, as there is no ambiguities there) would be best.

**Comment by samreid** [comment]
>  Instead, the remapping for filenames (but not directories, as there is no ambiguities there) would be best.

Can you please clarify by specifying the proposed filenames and directory names?

**Comment by pixelzoom** [comment]
My understanding what that the brand names were going to change as follows:

adapted-from-phet -> adaptedFromPhET
phet -> phet (no change)
phet-io -> phet-io

... and that this involved changing the subdirectory names in the brand repository, as well as the changes @jonathanolson mentioned in https://github.com/phetsims/chipper/issues/560#issuecomment-349545979.  I really think it's a bad idea to have (for example) "--brand=phet-io" that results in "phetio" appearing in directory and file names.

**Comment by samreid** [comment]
> phet-io -> phet-io

Did you mean phetio?

**Comment by pixelzoom** [comment]
Yep, fixed in https://github.com/phetsims/chipper/issues/560#issuecomment-349709083

**Comment by jonathanolson** [comment]
Changes required to change the brands (particularly phet-io to phetio, NOTE it's probably possible to choose some subset of this instead of all of it, but that removes most of the "benefits"):

- All code usages of `'phet-io` would presumably be changed to `'phetio'`, similar for filenames, directories, imports (`PHET_IO` => `PHETIO`), comments, etc.
  - There is in the range of ~2400 matches across our codebase for `phet-io`.
  - This doesn't count all of the potential renaming in server directories, server configs, etc.
- Query parameters for brands would change (`brand=phetio` instead, potentially breaking change unless we handle backwards compatibility)
- Rename repositories (phet-io, phet-io-website, phet-io-*)
  - This might cause a lot of broken links.
  - Any time we'd need to build an older (phet-io) branded simulation, we'd need to remap dependencies.json to change the repository names (a type of change we've never had to do before, as we haven't renamed any repositories after they were used for deployments)
- What do we do with existing phet-io sims?
- Presumably we keep phet-io.colorado.edu (so links don't break), or are we recommending changing to phetio.colorado.edu (new certificate, routing, etc.)
- Do we need to keep the branding, or will `PhET-iO` be changing? Presumably it stays the same, and we live with the inconsistency.
- Probably tons of things on the phet-io side that I'm not aware of. @samreid or @zepumph, does this break backwards compatibility or cause any major issues?
- Chipper/perennial build tools (this is probably minimal cost, already working on it). May need to add patches for backwards compatibility
- Will require reworking and full testing on:
  - Phet-io website/Apache (phet-io is baked into a lot of the directory names and configuration)
  - Yotta (will need to handle backwards compatibility with brands from sim messages)
  - Bayes continuous testing (has a lot of phet-io references, naming and specific behavior)
  - Phetmarks
  - Phettest.colorado.edu

Benefits received from the renaming:

- File version names are slightly easier to do/parse.
- We would have less of a mix of names in code, since `-` isn't allowed in variable names (still would probably keep some things, like phet-io.colorado.edu)

Instead, I'd propose we would ONLY change the suffix on built filenames, so e.g.:
```
faradays-law/

  1.0.0-dev.1/
    phet/
      faradays-law_en-phet.html
      faradays-law_all-phet.html
      faradays-law_all-phet-debug.html
    phet-io/
      faradays-law_all-phetio.html
      faradays-law_all-phetio-debug.html
```

If that is unacceptable, could we consider a different "separator" so that we could keep the brand names the same, e.g.:
```
faradays-law/

  1.0.0-dev.1/
    phet/
      faradays-law_en_phet.html
      faradays-law_all_phet.html
      faradays-law_all_phet_debug.html
    phetio/
      faradays-law_all_phet-io.html
      faradays-law_all_phet-io_debug.html
```
Since we tend to allow dashes in names (but not underscores), using the underscore as a separator works well.

**Comment by zepumph** [comment]
Tagging for dev meeting tomorrow.

**Comment by samreid** [comment]
Previously, @kathy-phet has expressed a strong preference for keeping the hyphen where PhET-iO appears in URLs, filenames and other user-visible locations.

**Comment by zepumph** [comment]
So by that we should probably have:
```
build/
    phet-io/
      faradays-law_all{{?}}phet-io.html
      faradays-law_all{{?}}phet-io{{?}}debug.html

**Comment by zepumph** [comment]
I'm not sure if this relates totally, but we should also try to factor out these names to a single place in chipper/perennial so there isn't a proliferation of assuming that the brand name won't change:
![image](https://user-images.githubusercontent.com/6856943/33688046-a2b4f676-da86-11e7-8ee5-a5e170021041.png)

**Comment by jonathanolson** [comment]
Also noting:

- What should versions be in the About dialog or anything that represents the running sim version? (Presumably it would include the brand always, even if it is -phet).
- Yotta was using the previous assumption that brands wouldn't use capital letters (among other things), so it will need to be upgraded anyways with whatever we decide.

**Comment by ariel-phet** [comment]
12/4/17 revisions to https://github.com/phetsims/chipper/issues/560#issuecomment-348304508 based on decision in https://github.com/phetsims/chipper/issues/560#issuecomment-349051862:

**dev directory structure** (assuming `brands=phet,phetio`)

```
faradays-law/

  1.0.0-dev.1/
    phet/
      faradays-law_en-phet.html
      faradays-law_all-phet.html
      faradays-law_all-phet-debug.html
    phetio/
      faradays-law_all-phetio.html
      faradays-law_all-phetio-debug.html

  1.0.0-rc.1/
    phet/
      faradays-law_en-phet.html
      faradays-law_all-phet.html
      faradays-law_all-phet-debug.html
    phetio/
      faradays-law_all-phetio.html
      faradays-law_all-phetio-debug.html

  1.0.0/
    phet/
      faradays-law_en-phet.html
      faradays-law_all-phet.html
      faradays-law_all-phet-debug.html
    phetio/
      faradays-law_all-phetio.html
      faradays-law_all-phetio-debug.html

  // one-off is in the directory name
  1.0.1-dev.1-sonification/
    phet/
      faradays-law_en-phet-sonification.html
      faradays-law_all-phet-sonification.html
      faradays-law_all-phet-sonification-debug.html
    phetio/
      faradays-law_all-phetio-sonification.html
      faradays-law_all-phetio-sonification-debug.html
```

**Comment by pixelzoom** [comment]
12/7/17 dev meeting notes, changes reflected in https://github.com/phetsims/chipper/issues/560#issuecomment-350054169:

• Allow dashes in brand name
• Keep "phet-io" and "adapted-from-phet" brand names
• Changed to underscore ('_') separator
• Continue to us underscore separator in locales (to separate language and country, e.g. "zh_CN")
• Version string (About dialog) and anything named 'version' in sim (e.g. globals) will not include the brand.

**Comment by mattpen** [comment]
Discussed with @jonathanolson.  We think that the build-server should fail and return an error if `adapted-from-phet` is in the list of brands.

**Comment by jonathanolson** [comment]
Implemented in the build and dev deploy. Will be waiting on the build-server request format and support for rc/production deploys.

See https://www.colorado.edu/physics/phet/dev/html/chains/1.7.0-dev.20 as an example.

**Comment by zepumph** [comment]
The commits above should make phet-io wrappers work for the updated version id schema.

**Comment by mattpen** [comment]
The build-server requests now accomodate the new version and filename scheme.  

@pixelzoom @jonathanolson @samreid - is it ok to close this issue?  Should we create documentation first?

**Comment by samreid** [comment]
I don't know.  What kind of documentation are you thinking of?

**Comment by jonathanolson** [comment]
Documentation (maybe in a perennial doc/*.md) about versioning is probably recommended.

**Comment by pixelzoom** [comment]
Can someone please summarize (in this issue) the format of the version id scheme that was actually implemented?

**Comment by jonathanolson** [comment]
> Can someone please summarize (in this issue) the format of the version id scheme that was actually implemented?

Notes in SimVersion.js:

```
 * The canonical description of our general versions:
 *
 * Each version string has the form: {{MAJOR}}.{{MINOR}}.{{MAINTENANCE}}[-{{TEST_TYPE}}.{{TEST_NUMBER}}] where:
 *
 * MAJOR: Sequential integer, starts at 1, and is generally incremented when there are significant changes to a simulation.
 * MINOR: Sequential integer, starts at 0, and is generally incremented when there are smaller changes to a simulation.
 *   Resets to 0 whenever the major number is incremented.
 * MAINTENANCE: Sequential integer, starts at 0, and is incremented whenever we build with the same major/minor (but with different SHAs).
 *   Resets to 0 whenever the minor number is incremented.
 * TEST_TYPE (when present): Indicates that this is a non-production build when present. Typically will take the values:
 *   'dev' - A normal dev deployment, which goes to spot (www.colorado.edu/physics/phet/dev/html/)
 *   'rc' -  A release-candidate deployment (off of a release branch). Also goes to spot only.
 *   anything else - A one-off deployment name, which is the same name as the branch it was deployed from.
 * TEST_NUMBER (when present): Indicates the version of the test/one-off type (gets incremented for every deployment).
 *   starts at 0 in package.json, but since it is incremented on every deploy, the first version published will be 1.
 *
 * It used to be (pre-chipper-2.0) that sometimes a shortened form of the (non-'phet') brand would be added to the end
 * (e.g. '1.3.0-dev.1-phetio' or '1.3.0-dev.1-adaptedfromphet'), or as a direct prefix for the TEST_TYPE (e.g.
 * 1.1.0-phetiodev.1 or 1.1.0-phetio). We have since moved to a deployment model where there are
 * subdirectories for each brand, so this is no longer part of the version. Since this was not used for any production sim
 * builds that we need statistics from, it is excluded in SimVersion.js or its description.
 *
 * Examples:
 *
 * 1.5.0 - Production simulation version (no test type). Major = 1, minor = 5, maintenance = 0
 * 1.5.0.rc.1 - Example of a release-candidate build version that would be published before '1.5.0' for testing.
 * 1.5.0.dev.1 - Example of a dev build that would be from master.
 * 1.5.0.sonification.1 - Example of a one-off build (which would be from the branch 'sonification')
```

**Comment by jonathanolson** [comment]
Closing, since this versioning has been fully implemented. If anyone needs anything else, please re-open.
